src/ports/linux/pnal_filetools.c

```diff
@@ -0,0 +1,202 @@
+
+/*********************************************************************
+ *        _       _         _
+ *  _ __ | |_  _ | |  __ _ | |__   ___
+ * | '__|| __|(_)| | / _` || '_ \ / __|
+ * | |   | |_  _ | || (_| || |_) |\__ \
+ * |_|    \__|(_)|_| \__,_||_.__/ |___/
+ *
+ * www.rt-labs.com
+ * Copyright 2020 rt-labs AB, Sweden.
+ *
+ * This software is dual-licensed under GPLv3 and a commercial
+ * license. See the file LICENSE.md distributed with this software for
+ * full license information.
+ ********************************************************************/
+
+#define _GNU_SOURCE /* Enable execvpe () */
+
+#include "pnal_filetools.h"
+
+#include "options.h"
+#include "osal_log.h"
+#include "pf_file.h"
+#include "pnal.h"
+
+#include <sys/stat.h>
+#include <sys/wait.h>
+
+#include <libgen.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+bool pnal_does_file_exist (const char * filepath)
+{
+   struct stat statbuffer;
+
+   return (stat (filepath, &statbuffer) == 0);
+}
+
+/**
+ * @internal
+ * Get the path to the directory where the main binary is located.
+ *
+ * @param path             Out:   Path to directory. Terminated string.
+ * @param size             In :   Size of the stringbuffer. Should be large
+ *                                enough for the full path to the binary.
+ *
+ * @return 0 on success.
+ *         -1 if an error occured.
+ */
+static int pnal_get_directory_of_binary (char * path, size_t size)
+{
+   ssize_t num_bytes = 0;
+
+   num_bytes = readlink ("/proc/self/exe", path, size);
+   if (num_bytes == -1)
+   {
+      return -1;
+   }
+
+   /* Verify that we have the full path (size is big enough) */
+   if (!pnal_does_file_exist (path))
+   {
+      return -1;
+   }
+
+   dirname (path);
```

Comment:
Doesn't dirname return the directory? From the man-page it _may_ modify the argument which is maybe what is happening here?

Reply:
Done